src/lib/device-logs/backends/loki.ts

```diff
@@ -0,0 +1,302 @@
+	createOrgIdMetadata,
+	Direction,
+	EntryAdapter,
+	PusherClient,
+	PushRequest,
+	PushResponse,
+	QuerierClient,
+	QueryRequest,
+	QueryResponse,
+	StreamAdapter,
+	TailRequest,
+	TailResponse,
+	Timestamp,
+} from 'loki-grpc-client';
+import { sbvrUtils } from '@resin/pinejs';
+import { LOKI_HOST, LOKI_PORT } from '../../../lib/config';
+import {
+	DeviceLog,
+	DeviceLogsBackend,
+	LogContext,
+	LogWriteContext,
+	Subscription,
+} from '../struct';
+import { captureException } from '../../../platform/errors';
+
+const { BadRequestError } = sbvrUtils;
+
+const MAX_RETRIES = 10;
+const MIN_BACKOFF = 100;
+const MAX_BACKOFF = 10 * 1000;
+const VERSION = 1;
+
+function sleep(ms: number) {
+	return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+function createTimestampFromDate(date = new Date()) {
+	const timestamp = new Timestamp();
+	timestamp.fromDate(date);
+	return timestamp;
+}
+
+function createTimestampFromMilliseconds(milliseconds: number) {
+	const MS_TO_NANOS = 1000000;
+	const seconds = Math.floor(milliseconds / 1000);
+	const nanos = (milliseconds - seconds * 1000) * MS_TO_NANOS;
+
+	const timestamp = new Timestamp();
+	timestamp.setSeconds(seconds);
+	timestamp.setNanos(nanos);
+	return timestamp;
+}
+
+export class LokiBackend implements DeviceLogsBackend {
+	private subscriptions: EventEmitter;
+	private querier: QuerierClient;
+	private pusher: PusherClient;
+	private ready: boolean;
+	private tailCalls: Map<string, grpc.ClientReadableStream<TailResponse>>;
+
+	constructor() {
+		this.subscriptions = new EventEmitter();
+		this.querier = new QuerierClient(
+			`${LOKI_HOST}:${LOKI_PORT}`,
+			createInsecureCredentials(),
+		);
+		this.pusher = new PusherClient(
+			`${LOKI_HOST}:${LOKI_PORT}`,
+			createInsecureCredentials(),
+		);
+		this.ready = false;
+		this.tailCalls = new Map();
+	}
+
+	public get available(): boolean {
+		return !this.ready;
+	}
+
+	/**
+	 *
+	 * Return $count of logs matching device_id in descending (BACKWARD) order.
+	 *
+	 * The logs are sorted by timestamp since Loki returns a distinct stream for each label combination.
+	 *
+	 * @param ctx
+	 * @param count
+	 */
+	public history(ctx: LogContext, count: number): Promise<DeviceLog[]> {
+		return new Promise((resolve, reject) => {
+			const oneHourAgo = new Date(Date.now() - 10000 * 60);
+
+			const queryRequest = new QueryRequest();
+			queryRequest.setSelector(this.getDeviceQuery(ctx));
+			queryRequest.setLimit(Number.isFinite(count) ? count : 1000);
+			queryRequest.setStart(createTimestampFromDate(oneHourAgo));
+			queryRequest.setEnd(createTimestampFromDate());
+			queryRequest.setDirection(Direction.BACKWARD);
+
+			const deviceLogs: DeviceLog[] = [];
+			const call = this.querier.query(
+				queryRequest,
+				createOrgIdMetadata(String(ctx.belongs_to__application)),
+			);
+			call.on('data', (queryResponse: QueryResponse) => {
+				deviceLogs.push(
+					...this.fromStreamsToDeviceLogs(queryResponse.getStreamsList()),
+				);
```

Comment:
I think it would make sense to do
```suggestion
				streamLogs.push(
					...queryResponse.getStreamsList(),
				);
```
and then
```ts
const streamLogs = await new Promise(...);
return _.orderBy(this.fromStreamsToDeviceLogs(deviceLogs), 'timestamp', 'desc');
```
so that we only process the logs if we need to, since if an error is hit partway through then we can skip the work. The counterpoints are:
1. if the processing would potentially lock up the event loop, but if that's possible then I think we'd need to put in dedicated measures regardless of when we process
2. if the success latency is particularly important to minimize at the cost of potentially wasted work

Reply:
Ok makes sense, updated.